src/tests/test_types.cpp

```diff
@@ -263,5 +263,105 @@ INSTANTIATE_TEST_SUITE_P(TestStringSuite, TestString,
+namespace {
+using DatePtr = std::unique_ptr<std::byte[], mo::DateTypeDeleter>;
 
+DatePtr DatePtrFromString(std::string const& s, std::unique_ptr<mo::DateType> const& date_type) {
+    return {date_type->MakeValue(boost::gregorian::from_simple_string(s)),
+            date_type->GetDeleater()};
+}
+}  // namespace
+
+TEST(TestDateSuite, Compare) {
+    std::unique_ptr<mo::DateType> date_type(
+            mo::CreateSpecificType<mo::DateType>(mo::TypeId::kDate, true));
+    auto test = [date_type = std::move(date_type)](
+                        std::string const& date1, std::string const& date2, mo::CompareResult res) {
+        DatePtr ptr1 = DatePtrFromString(date1, date_type);
+        DatePtr ptr2 = DatePtrFromString(date2, date_type);
+        EXPECT_EQ(date_type->Compare(ptr1.get(), ptr2.get()), res);
+    };
+    test("2004-07-28", "2004-07-28", mo::CompareResult::kEqual);
+    test("2005-01-20", "2004-07-28", mo::CompareResult::kGreater);
+    test("2023-09-01", "2024-10-25", mo::CompareResult::kLess);
+    test("2023-10-28", "2023-10-28", mo::CompareResult::kEqual);
+    test("2203-09-01", "2024-10-25", mo::CompareResult::kGreater);
+    test("1991-11-20", "2200-07-28", mo::CompareResult::kLess);
+}
+
+TEST(TestDateSuite, Dist) {
+    std::unique_ptr<mo::DateType> date_type(
+            mo::CreateSpecificType<mo::DateType>(mo::TypeId::kDate, true));
+    auto test = [date_type = std::move(date_type)](std::string const& date1,
+                                                   std::string const& date2, long res) {
+        DatePtr left_val = DatePtrFromString(date1, date_type);
+        DatePtr right_val = DatePtrFromString(date2, date_type);
+        EXPECT_EQ(static_cast<long>(date_type->Dist(left_val.get(), right_val.get())), res);
+    };
+    test("2004-07-28", "2004-07-28", 0);
+    test("2005-01-20", "2004-07-28", 176);
+    test("1991-11-20", "2200-07-28", 76221);
+    test("1941-06-22", "1945-05-09", 1417);
+    test("2023-10-28", "2297-08-12", 100000);
+}
+
+namespace {
+struct TestDateArithmeticsParam {
+    std::string const date1;
+    std::string const date2;
+    long const dist;
+    TestDateArithmeticsParam(std::string l, std::string r, long dist)
+        : date1(std::move(l)), date2(std::move(r)), dist(dist){};
+};
+
+void TestBinop(std::string const& initial_date_, std::string const& expected_, long const delta_,
+               std::byte* (mo::DateType::*Binop)(std::byte const* date, mo::DateType::Delta const&)
+                       const) {
+    std::unique_ptr<mo::DateType> date_type(
+            mo::CreateSpecificType<mo::DateType>(mo::TypeId::kDate, true));
+    mo::DateType::Delta delta(delta_);
+
+    DatePtr initial_date = DatePtrFromString(initial_date_, date_type);
+    DatePtr expected = DatePtrFromString(expected_, date_type);
+    auto actual =
+            DatePtr((date_type.get()->*Binop)(initial_date.get(), delta), date_type->GetDeleater());
```

Comment:
```suggestion
            DatePtr(std::invoke(Binop, date_type.get(), initial_date.get(), delta), date_type->GetDeleater());
```
Method pointer call syntax looks atrocious, let's use `std::invoke`